File: internal/model/echo/echo.go
package model

import (
	"time"

	fileModel "github.com/lin-snow/ech0/internal/model/file"
	uuidUtil "github.com/lin-snow/ech0/internal/util/uuid"
	"gorm.io/gorm"
)

// Echo 定义Echo实体
type Echo struct {
	ID            string               `gorm:"type:char(36);primaryKey"                      json:"id"`
	Content       string               `gorm:"type:text;not null"                            json:"content"`
	Username      string               `gorm:"type:varchar(100)"                             json:"username,omitempty"`
	EchoFiles     []fileModel.EchoFile `gorm:"foreignKey:EchoID;constraint:OnDelete:CASCADE" json:"echo_files,omitempty"`
	Layout        string               `gorm:"type:varchar(50);default:'waterfall'"          json:"layout,omitempty"`
	Private       bool                 `gorm:"default:false;index:idx_echos_private_created,priority:1" json:"private"`
	UserID        string               `gorm:"type:char(36);not null;index"                  json:"user_id"`
	Extension     *EchoExtension       `gorm:"foreignKey:EchoID;constraint:OnDelete:CASCADE" json:"extension,omitempty"`
	Tags          []Tag                `gorm:"many2many:echo_tags;"                          json:"tags,omitempty"`
	FavCount      int                  `gorm:"default:0"                                     json:"fav_count"`
	CreatedAt     time.Time            `gorm:"index:idx_echos_private_created,priority:2"    json:"created_at"`
}

type EchoExtension struct {
	ID        string                 `gorm:"type:char(36);primaryKey"      json:"id"`
	EchoID    string                 `gorm:"type:char(36);not null;uniqueIndex" json:"echo_id"`
	Type      string                 `gorm:"type:varchar(100);not null"    json:"type"`
	Payload   map[string]interface{} `gorm:"serializer:json;type:text;not null" json:"payload"`
	CreatedAt time.Time              `json:"created_at"`
	UpdatedAt time.Time              `json:"updated_at"`
}

// Tag 定义Tag实体
type Tag struct {
	ID         string    `gorm:"type:char(36);primaryKey"              json:"id"`
	Name       string    `gorm:"type:varchar(50);uniqueIndex;not null" json:"name"`
	UsageCount int       `gorm:"default:0"                             json:"usage_count"`
	CreatedAt  time.Time `                                             json:"created_at"`
}

// EchoTag 纯关系表，联合主键
type EchoTag struct {
	EchoID string `gorm:"type:char(36);primaryKey"`
	TagID  string `gorm:"type:char(36);primaryKey;index"`
}

func (e *Echo) BeforeCreate(_ *gorm.DB) error {
	if e.ID == "" {
		e.ID = uuidUtil.MustNewV7()
	}
	return nil
}

func (t *Tag) BeforeCreate(_ *gorm.DB) error {
	if t.ID == "" {
		t.ID = uuidUtil.MustNewV7()
	}
	return nil
}

func (e *EchoExtension) BeforeCreate(_ *gorm.DB) error {
	if e.ID == "" {
		e.ID = uuidUtil.MustNewV7()
	}
	return nil
}

const (
	Extension_MUSIC      = "MUSIC"
	Extension_VIDEO      = "VIDEO"
	Extension_GITHUBPROJ = "GITHUBPROJ"
	Extension_WEBSITE    = "WEBSITE"

	LayoutWaterfall  = "waterfall"
	LayoutGrid       = "grid"
	LayoutHorizontal = "horizontal"
	LayoutCarousel   = "carousel"
)

File: hub/src/types/echo.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025-2026 lin-snow

/** 与后端 Echo 模型对齐（见 internal/model/echo） */
export interface EchoTag {
  id: string
  name: string
  /** 原始可能为 ISO 文本；经 `queryInstancePage` 后为 Unix 秒（number） */
  created_at?: number | string
  usage_count?: number
}

/** 帖子正文与附件（Hub 聚合列表接口返回的字段） */
export interface EchoPost {
  id: string
  content: string
  username?: string
  /**
   * 原始可能为 ISO 文本或 Unix（秒/毫秒）；经 `queryInstancePage` 后为 Unix 秒（number）。
   */
  created_at: number | string
  fav_count?: number
  tags?: EchoTag[]
  echo_files?: App.Api.Ech0.EchoFile[]
  layout?: string
  extension?: App.Api.Ech0.EchoExtension | null
  private?: boolean
  user_id?: string
}

export interface HubPostMeta {
  instanceId: string
  instanceUrl: string
}

export type EchoPostWithHub = EchoPost & { _hub: HubPostMeta }

/** 统一 API 包装 internal/model/common/result.go */
export interface ApiResult<T> {
  code: number
  msg: string
  data: T
}

export interface EchoQueryPage {
  total: number
  items: EchoPost[]
}
